pkg/api/validation/validation.go

```diff
@@ -1594,6 +1594,23 @@ func validateTolerations(tolerations []api.Toleration, fldPath *field.Path) fiel
 		// validate the toleration key
 		allErrors = append(allErrors, unversionedvalidation.ValidateLabelName(toleration.Key, idxPath.Child("key"))...)
 
+		// validate forgiveness toleration
+		if toleration.Key == unversioned.TaintNodeDown {
```

Comment:
I think you should refactor this block a part that is specific to `toleration.Key == unversioned.TaintNodeDown`
and a part that applies to all ForgivenessSeconds. In other words, I think you should change it to something like this
```
if toleration.ForgivenessSeconds != nil {
    if toleration.Effect != api.TaintEffectNoExecute {
         allErrors = append(allErrors, field.Invalid(idxPath.Child("effect"), toleration.Effect, "Effect must be NoExecute when `ForgivenessSeconds` is set"))
    }
    if *toleration.ForgivenessSeconds <= 0 {
        allErrors = append(allErrors, field.Invalid(idxPath.Child("forgivenessSeconds"), toleration.ForgivenessSeconds, "forgivenessSeconds must be greater than zero when set"))
    }
    // TODO: Remove all code from here to the end of the enclosing block once ForgivenessSeconds is fully supported
    if toleration.Key != unversioned.TaintNodeDown {
       // an error saying that currently only NodeDown toleration supports ForgivenessSeconds
    }
    if toleration.Operator != api.TolerationOpExists {
       // same code as you have now for this case
    }
  }
```

You don't need the check for `len(toleration.Value) > 0` because there is already code later in this function that does exactly the same thing.